### DISCUSSION TITLE
Possible for IDE inter-module resolution to ignore the version ?

### DISCUSSION THREAD

**Discussion by fmarot** [eclipse-m2e/m2e-core#1008] [Q&A]
Hello all,
suppose I have in Eclipse 2 projects: A v1 and B v2. But A's pom references B v1.
By default, Eclipse will put the local repo's jar (in $HOME/.m2/) in A's classpath instead of directly referencing B because the version differ.
Is there any setting that would allow all modules to be referenced if they are available in the IDE independently of the version ?

Why am I asking this ? Because I want to have fully dynamic versions in my pom. So A will have in its pom `<version>${revision}${changelist}</version>` but it will reference B with `<version>1</version>`. But locally, B's version will also be `<version>${revision}${changelist}</version>` in its pom ! Not `<version>1</version>`.
The goal is to have multiple inter-dependant projects being runnable/debuggable in the IDE while being CI-friendly (the CI will set revision to a number and changelist to the git branch name).

Happy Halloween all !
François

**Comment by laeubi** [comment]
I think it definatly would help to share an example, CI-Friendly Versions are mostly not supported very well in m2e but we might can improve support in that aread. I assume you have enabled resoloution of workspace artifacts already?

**Reply by fmarot** [reply]
I'll try to share a sample if I wasn't clear, sorry.
Regarding workspace artifact resolution, I think it's activated by default for compilation and when running Eclipse's .launch files, isn't it ?
Because it works as expected out of the box (when projet A v1 's pom references project B v1 and B v1 is also in the workspace, the link is correctly established).

**Reply by HannesWell** [reply]
Once PR https://github.com/eclipse-m2e/m2e-core/pull/852 is completed, CI-friendly versions should work in M2E/the IDE as well.

**Reply by fmarot** [reply]
Here is a visual attempt at explaining myself:

![workspace-classpath](https://user-images.githubusercontent.com/633019/198374158-42cedcb7-6935-446d-9d74-b5338e911c92.png)

![local-repo-jar-classpath](https://user-images.githubusercontent.com/633019/198374201-e395c006-df10-4eae-b66b-cb61c8fdea38.png)

do you think it could be made to work ?

**Reply by mickaelistria** [reply]
How does Maven behave in that case? Suppose you have both bbb:2 and aaa:1 (referencing bbb:1) in your reactor; does Maven reactor reference bbb:2 when resolving aaa:1?
As m2e uses the same resolver, it should have similar behavior -at least that's what one would expect-; so it could be a bug or a limitation.
Ideally, if you can create a test-case for it and contribute it as a PR to the org.eclipse.m2e.core.tests repo, it will be very possible to at least try fixing it.

**Reply by lalmeras** [reply]
Maven is strict about dependencies. I don't think m2e should permit something maven do not allow. And as stated by @mickaelistria, I don't think it to be possible.

From a pure maven point of view, I think there is no simple solution to this, except using a property for B dependency version, set at build time.

**Reply by laeubi** [reply]
@fmarot have you tried using `<version>1,2</version>` ?

**Reply by fmarot** [reply]
(first of all sorry for the delay in responding, and thanks for your time and responses)

> How does Maven behave in that case? Suppose you have both bbb:2 and aaa:1 (referencing bbb:1) in your reactor; does Maven reactor reference bbb:2 when resolving aaa:1?

Yes, when run from command line and in the same reactor, Maven exhibits the same behavior as Eclipse/m2e: the version used is the one from the reactor if it maches, from the local repo otherwise.

I oversimplified the description of what I want to explain here.
In fact in reality, on the command-line I'm talking about 2 different maven multi-module projects compiled with 2 different Maven execution.
But I import those 2 different projects in the same IDE workspace. So on the command line it works but the conditions are different than in the IDE. Due to CIFriendly workarounds (Jenkins sets the version dynamically through -Dchangelist=branchname[-SNAPSHOT] at build-time), m2e does not know the 'real' version of each pom. But the <dependencies> versions are hardcoded, while devs would just want m2e to use whatever version/branch they chose to have in their IDE. I would expect it to 'just' work, while on the command line it is required to pass the version in order to build.

Sorry, I imagine it's not very clear...

> @fmarot have you tried using `<version>1,2</version>` ?

sorry, I did not tried yet. But I'd prefer avoid that because while it would (maybe) solve the problem in the IDE, it would induce incertainty during the build on jenkins.
